### ISSUE TITLE
Signed zero

### ISSUE THREAD

**Issue by fredrik-johansson** [mpmath/mpmath#167]
bc.. Should negative zero (-0) be supported? It is virtually supported already;
the object 'fnzero' is defined in libmpf.py (though not used anywhere). It
just needs to be supported in standard operations, and elsewhere "x ==
fzero" needs to be replaced with something else so that -0 is not forwarded
to places where it shouldn't go.

I'm not personally convinced that signed zero is worth the trouble, but it
would be useful for allowing extended compatibility with IEEE 754
semantics, for continuity along branch cuts, etc, and could resolve some
issues in interval arithmetic (although it would also cause some more
complication).

p. Original issue for "#167":https://github.com/fredrik-johansson/mpmath/issues/167: "http://code.google.com/p/mpmath/issues/detail?id=127":http://code.google.com/p/mpmath/issues/detail?id=127

p. Original author: "https://code.google.com/u/111502149103757882156/":https://code.google.com/u/111502149103757882156/

p. Original owner: "https://code.google.com/u/111502149103757882156/":https://code.google.com/u/111502149103757882156/

**Comment by skirpichev** [comment]
@fredrik-johansson, I'm not sure why this issue was closed.  Could you please explain why?   Do you think there is support for signed zero?

I don't think it's even can be created with a public API:
```
In [100]: mpmath.mpf(0.0)
Out[100]: mpf('0.0')

In [101]: mpmath.mpf(-0.0)
Out[101]: mpf('0.0')

In [102]: mpmath.mpf((0,0,0,0))
Out[102]: mpf('0.0')

In [103]: mpmath.mpf((1,0,0,0))
Out[103]: mpf('0.0')

In [104]: mpmath.mpf((1,0,0,0))._mpf_
Out[104]: (0, 0, 0, 0)
```

See https://github.com/sympy/sympy/issues/25486, where it's a real issue.

**Comment by WarrenWeckesser** [comment]
FYI: `complex` also accepts two arguments, the real and imaginay parts:

```
In [1]: complex(1, -0.0)
Out[1]: (1-0j)
```

**Comment by achan001** [comment]
```
In [12]: 1/(-2+0j)
Out[12]: (-0.5-0j)

In [13]: 1/(-2-0j)
Out[13]: (-0.5-0j)
```

This is not a bug.

Complex division does not just multiply by conjugates.
It does scaling first, to avoid intermediate overflow.

(1,0) / (-2,±0) = -0.5 * (1,0)/(1,∓0) * (1,±0)/(1,±0)

num = -0.5 * (1,0) * (1,±0) = -0.5 * (1,[0±0](https://en.wikipedia.org/wiki/Signed_zero#Arithmetic)) = -0.5 * (1,0) = (-0.5,-0)
den = (1,∓0) * (1,±0) = 1^2 + 0^2 = 1

Of course, we can scale numerator differently.

num = 0.5 * (-1,**−0**) * (1,±0) = 0.5 * (-1,[**−0**∓0](https://en.wikipedia.org/wiki/Signed_zero#Arithmetic)) = 0.5 * (-1,∓0) = (-0.5,∓0)

It cost extra logic to make sure numerator imag part is negative.
I don't know if this is worth it.

**Comment by skirpichev** [comment]
> This is not a bug.

That depends...  Most people, probably, recognize ``-2-0j`` as a rectangular notation (i.e. as an equivalent of ``complex(-2, -0.0)``), while in the Python this actually means ``complex(-2, 0) - complex(0.0,0.0) = complex(-2, 0)``.

See https://discuss.python.org/t/33433

**Comment by achan001** [comment]
The issue is numerator 1 get promoted to complex (1+0j)
Signed zero does not mean exactly zero, just infinitesimally small.

 (1+0j) ≡ 1∠(ε), phase angle not exactly zero, but tiny non-negative ε
-(1+0j) ≡ 1∠(ε - pi), phase angle not exactly -pi either

(1+0j) / (-2-0j) ≡ (1∠(ε1)) / (2∠(ε2-pi)) ≡ 0.5 ∠((ε1-ε2) + pi)

Result is ambiguous, if sign(ε1-ε2) = 1, phase angle flip to -pi

(1+0j) / (-2-0j) ≡ 0.5 ∠(±pi) = -0.5 ± 0j

**Comment by achan001** [comment]
The other case is more clear-cut

(1+0j) / (-2+0j) ≡ (1∠(ε1)) / (2∠(pi-ε2)) ≡ 0.5 ∠((ε1+ε2) - pi)

(1+0j) / (-2+0j) = 0.5 ∠(-pi) = -0.5 - 0j

**Comment by achan001** [comment]
> The issue is numerator 1 get promoted to complex (1+0j)

This is the reason we have (1±0j) * 1 = (1+0j)

**Comment by skirpichev** [comment]
> This is the reason we have (1±0j) * 1 = (1+0j)

No.  The reason is that ``1-0j=complex(1,0)-complex(0,0)=1+0j``.  To get ``complex(1, -0.0)`` with imaginary literals you should use cumbersome syntax ``-(-1+0j)``.  See mentioned discussion.

**Comment by skirpichev** [comment]
Here is a draft pr: #768

**Comment by skirpichev** [comment]
Diofant tests on the above patch: https://github.com/diofant/diofant/pull/1391

**Comment by skirpichev** [comment]
Well, I'm going to close this issue as not planned.

Support for signed zero will break backward compatibility in handling branch cut, see e.g. [this](https://github.com/mpmath/mpmath/issues/865#issuecomment-2345564168).  Probably it's too much for projects like SymPy.

We also can simplify mpf implementation.
